src/logic/indexed_eq.lean

```diff
@@ -0,0 +1,74 @@
+/-
+Copyright (c) 2021 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser
+-/
+import logic.basic
+
+/-!
+# Indexed equality
+
+This file provides `indeq` with notation `a =ᵢ b`, which provide a different notion of heterogeneous
+equality to `heq` (`a == b`).
+
+In particular, `indeq` is for when `a : A i` and `b : A j`, where `A : ι → Sort u` is an indexed
+family of types. Instead of `type_eq_of_heq` which states `A i = A j` for `a == b`, this provides
+`indeq.index_eq` which states `i = j` for `a =ᵢ b`.begin
+
+Note that in the degerate case when `A = id`, this is equivalent to `heq`.
+-/
+
+universes ui ui' u u'
+
+section
+
+variables {ι : Sort ui} {ι' : Sort ui'} {A : ι → Sort u} {A' : ι' → Sort u'}
+
+/-- Indexed equality; `(ai : A i) =ᵢ (aj : A j)` if `i = j` -/
+inductive indeq {i : ι} (ai : A i) : Π {j} (aj : A j), Prop
+| refl [] : indeq ai
+
+infix ` =ᵢ `:50 := indeq
+
+attribute [refl] indeq.refl
+
+variables {i j k : ι} {a a' : A i} {b : A j} {c : A k}
+
+
+@[symm] lemma indeq.symm (h : a =ᵢ b) : b =ᵢ a :=
+indeq.rec_on h (indeq.refl a)
+
+@[trans] lemma indeq.trans (h₁ : a =ᵢ b) (h₂ : b =ᵢ c) : a =ᵢ c :=
+indeq.rec_on h₂ h₁
+
+lemma indeq_of_eq (h : a = a') : a =ᵢ a' :=
```

Comment:
How can Lean infer `A` in this lemma? I suspect it whould be good making it explicit.